x/records/keeper/callback_transfer_test.go

```diff
@@ -0,0 +1,126 @@
+		DepositEpochNumber: 1,
+		HostZoneId:         chainId,
+		Amount:             balanceToStake,
+		Status:             recordtypes.DepositRecord_TRANSFER,
+	}
+	s.App.RecordsKeeper.SetDepositRecord(s.Ctx(), depositRecord)
+	packet := channeltypes.Packet{Data: s.MarshalledICS20PacketData()}
+	ack := s.ICS20PacketAcknowledgement()
+	callbackArgs := types.TransferCallback{
+		DepositRecordId: depositRecord.Id,
+	}
+	args, err := s.App.RecordsKeeper.MarshalTransferCallbackArgs(s.Ctx(), callbackArgs)
+	s.Require().NoError(err)
+
+	return TransferCallbackTestCase{
+		initialState: TransferCallbackState{
+			callbackArgs: callbackArgs,
+		},
+		validArgs: TransferCallbackArgs{
+			packet: packet,
+			ack:    &ack,
+			args:   args,
+		},
+	}
+}
+
+func (s *KeeperTestSuite) TestTransferCallback_Successful() {
+	tc := s.SetupTransferCallback()
+	initialState := tc.initialState
+	validArgs := tc.validArgs
+
+	err := recordskeeper.TransferCallback(s.App.RecordsKeeper, s.Ctx(), validArgs.packet, validArgs.ack, validArgs.args)
+	s.Require().NoError(err)
+
+	// Confirm deposit record has been updated to STAKE
+	record, found := s.App.RecordsKeeper.GetDepositRecord(s.Ctx(), initialState.callbackArgs.DepositRecordId)
+	s.Require().True(found)
+	s.Require().Equal(record.Status, recordtypes.DepositRecord_STAKE, "deposit record status should be STAKE")
+}
+
+func (s *KeeperTestSuite) checkTransferStateIfCallbackFailed(tc TransferCallbackTestCase) {
+	record, found := s.App.RecordsKeeper.GetDepositRecord(s.Ctx(), tc.initialState.callbackArgs.DepositRecordId)
+	s.Require().True(found)
+	s.Require().Equal(record.Status, recordtypes.DepositRecord_TRANSFER, "deposit record status should be TRANSFER")
+}
+
+func (s *KeeperTestSuite) TestTransferCallback_TransferCallbackTimeout() {
+	tc := s.SetupTransferCallback()
+	invalidArgs := tc.validArgs
+	// a nil ack means the request timed out
+	invalidArgs.ack = nil
+	err := recordskeeper.TransferCallback(s.App.RecordsKeeper, s.Ctx(), invalidArgs.packet, invalidArgs.ack, invalidArgs.args)
+	s.Require().NoError(err)
+	s.checkTransferStateIfCallbackFailed(tc)
+}
+
+func (s *KeeperTestSuite) TestTransferCallback_TransferCallbackErrorOnHost() {
+	tc := s.SetupTransferCallback()
+	invalidArgs := tc.validArgs
+	// an error ack means the tx failed on the host
+	errorAck := channeltypes.Acknowledgement{Response: &channeltypes.Acknowledgement_Error{Error: "error"}}
+
+	err := recordskeeper.TransferCallback(s.App.RecordsKeeper, s.Ctx(), invalidArgs.packet, &errorAck, invalidArgs.args)
+	s.Require().EqualError(err, "TransferCallback does not handle errors: error: invalid request")
+	s.checkTransferStateIfCallbackFailed(tc)
+}
+
+func (s *KeeperTestSuite) TestTransferCallback_WrongCallbackArgs() {
+	tc := s.SetupTransferCallback()
+	invalidArgs := tc.validArgs
+
+	err := recordskeeper.TransferCallback(s.App.RecordsKeeper, s.Ctx(), invalidArgs.packet, invalidArgs.ack, []byte("random bytes"))
+	s.Require().EqualError(err, "unexpected EOF")
```

Comment:
wrap err in original function to make this error little less cryptic. We have a `ErrUnmarshalFailure` error type I think (although it might be in stakeibc)

Reply:
done